### ISSUE TITLE
Custom rule jar/class not found by CheckStyle

### ISSUE THREAD

**Issue by SteffeC** [checkstyle/checkstyle#712]
Hi,

-I'm trying to create a custom rule below the TreeWalker module in CheckStyle.

-I'm following the example in http://checkstyle.sourceforge.net/writingchecks.html but can't make CheckStyle find my custom rule class.

-Using the command in http://checkstyle.sourceforge.net/writingchecks.html#Integrate_your_check :

C:\>java -classpath my_rule.jar:checkstyle-6.4-all.jar -c config.xml -r .  (my_rule.jar contains my rule class, config.xml is a subset of the modules in \checkstyle-6.4-src\checkstyle-6.4\config\checkstyle_checks.xml)

-Gives the error below:

Unrecognized option: -c
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

-If adding "-jar checkstyle-6.4-all.jar" to the command:

C:\>java -classpath my_rule.jar:checkstyle-6.4-all.jar -jar checkstyle-6.4-all.jar -c config.xml -r .

-It gives another error:

org.apache.commons.cli.UnrecognizedOptionException: Unrecognized option: -r
        at org.apache.commons.cli.Parser.processOption(Parser.java:363)
        at org.apache.commons.cli.Parser.parse(Parser.java:199)
        at org.apache.commons.cli.Parser.parse(Parser.java:85)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:77)
usage: java com.puppycrawl.tools.checkstyle.Main [options] -c <config.xml>
            file...
 -c <arg>   The check configuration file to use.
 -f <arg>   Sets the output format. (plain|xml). Defaults to plain
 -o <arg>   Sets the output file. Defaults to stdout
 -p <arg>   Loads the properties file
 -v         Print product version and exit

-If removing "-r" from the the command:

C:\>java -classpath my_rule.jar:checkstyle-6.4-all.jar -jar checkstyle-6.4-all.jar -c config.xml .

-The default check works OK:

Starting audit...
C:\.\Test.java:472:17: Name 'TestInvalidMethodName' must match pattern '^[a-z][a-zA-Z0-9]*$'.
C:\.\Test.java:499:17: Name 'TestAllErrorsInCombination' must match pattern'^[a-z][a-zA-Z0-9]*$'.
Audit done.

-The default check also works with the two alternative commands below:

C:\>java -classpath checkstyle-6.4-all.jar -jar checkstyle-6.4-all.jar -c config.xml .  (i.e. removed my_rule.jar from classpath)
C:\>java -jar checkstyle-6.4-all.jar -c config.xml .                                    (i.e. removed all jars from classpath)

**** MY PROBLEM IS DESCRIBED BELOW ***

-My custom rule check doesn't work with the command below:

C:\>java -classpath my_rule.jar:checkstyle-6.4-all.jar -jar checkstyle-6.4-all.jar -c my_config.xml . (my_config.xml contains same modules as in config.xml plus my custom rule TestCaseAnnotationCheck class)

-It looks like the TestCaseAnnotationCheck class is not found on the classpath: 

Unable to create Checker: cannot initialize module TreeWalker - Unable to instantiate com.my.checkstyle.rule.TestCaseAnnotationCheck
com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - Unable to instantiate com.my.checkstyle.rule.TestCaseAnnotationCheck
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:184)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:185)
        at com.puppycrawl.tools.checkstyle.Main.createChecker(Main.java:150)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:127)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to instantiate com.my.checkstyle.rule.TestCaseAnnotationCheck
        at com.puppycrawl.tools.checkstyle.PackageObjectFactory.createModule(PackageObjectFactory.java:156)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:182)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:185)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:163)
        ... 3 more
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to instantiate com.my.checkstyle.rule.TestCaseAnnotationCheckCheck
        at com.puppycrawl.tools.checkstyle.PackageObjectFactory.doMakeObject(PackageObjectFactory.java:98)
        at com.puppycrawl.tools.checkstyle.PackageObjectFactory.createModule(PackageObjectFactory.java:153)
        ... 6 more

-Why is "Check" appended on the TestCaseAnnotationCheck class name on this line?
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to instantiate com.my.checkstyle.rule.TestCaseAnnotationCheckCheck 

-I get the same error when removing my_rule.jar from the classpath in the command:

C:\>java -jar checkstyle-6.4-all.jar -c my_config.xml .

-It's like CheckStyle can't see my rule jar/class?

-How do I make CheckStyle find my rule? :)

-NOTE: My TestCaseAnnotationCheck rule class and my_config.xml works when running it with CheckStyle in Eclipse.

Regards
SteffeC

**Comment by rdiachenko** [comment]
Hi @SteffeC 

Which operating system do you use?

Probably you missed this:
>> Then run (with the path separator adjusted to your platform):

**Comment by romani** [comment]
Did you do "Integrate Your Check" from http://checkstyle.sourceforge.net/writingchecks.html

Where is no "-r" option from 6.3

**Comment by SteffeC** [comment]
I'm working on Windows so I should have used ';' as separator instead of ':'.

But the main problem was that I used "-jar checkstyle-6.4-all.jar" to point out the main method, because -jar makes Java ignore the -classpath arguments.

So instead I used "com.puppycrawl.tools.checkstyle.Main" to point out the CheckStyle main method.

This command works for me on Windows:
java -classpath my_rule.jar;checkstyle-6.4-all.jar com.puppycrawl.tools.checkstyle.Main -c my_config.xml .

I suggest that you update the existing command example on http://checkstyle.sourceforge.net/writingchecks.html#Integrate_your_check from:
java -classpath mycompanychecks.jar:checkstyle-6.4-all.jar \ -c config.xml -r .

To:
java -classpath mycompanychecks.jar:checkstyle-6.4-all.jar com.puppycrawl.tools.checkstyle.Main -c config.xml .  (i.e. add Main class and remove -r argument since this is no longer supported)

Thanks for your help!
/SteffeC
